sw/device/tests/sram_ctrl_execution_test_ret.c

```diff
@@ -0,0 +1,175 @@
+     *   |  0000000  |  01010  |  01011  |  000  | 01010  |  0110011  |
+     *   +-----------+---------+---------+-------+--------+-----------+
+     *   31        25 24     20 19     15 14   12 11     7 6          0
+     */
+    0x00A58533,
+    /*
+     *   Abstract representation: ret == jalr rd, offset(rs1)
+     *   +----------------+-------+-------+------+-----------+
+     *   |  000000000000  |  rs1  |  000  |  rd  |  1100111  |
+     *   +----------------+-------+-------+------+-----------+
+     *   31             20 19   15 14   12 11   7 6          0
+     *
+     *   Custom representation: jalr zero, 0(ra)
+     *   +----------------+---------+-------+---------+-----------+
+     *   |  000000000000  |  00001  |  000  |  00000  |  1100111  |
+     *   +----------------+---------+-------+---------+-----------+
+     *   31             20 19     15 14   12 11      7 6          0
+     */
+    0x00008067,
+};
+
+/*
+ * Handles an exception.
+ *
+ * This exception handler keeps default exception implementations for faults
+ * that are not relevant in terms of this test. The only fault handler that
+ * is overridden is Instruction Access Fault.
+ *
+ * The controlled Instruction Access Fault occurs originates in the retention
+ * SRAM, which means that normally the return address would be calculated
+ * relevant to the offending instruction. However, due to execution from
+ * retention SRAM being permanently disabled, this approach would not work.
+ * The return address for the one stack frame before is used instead. In
+ * other words sram_execution_test -> retention_sram -> exception_handler
+ * -> sram_execution_test.
+ *
+ * Before the jump into the exception handler, the register set is saved on
+ * stack. This means that the return address can be loaded from the stack, or
+ * simply read from the return address register (before it is overridden by
+ * the code in this exception handler). This means that the return address
+ * register value must be read before any function call.
+ */
+void handler_exception(void) {
+  // Save the return address register value.
+  uint32_t ret_addr = 0;
+  asm volatile("sw ra, %0" : "=m"(ret_addr));
+
+  uint32_t mcause;
+  exc_id_t exc_cause;
+
+  CSR_READ(CSR_REG_MCAUSE, &mcause);
+  exc_cause = (exc_id_t)(mcause & kIdMax);
+
+  switch (exc_cause) {
+    case kInstMisa:
+      handler_instr_acc_fault();
+      break;
+    case kInstAccFault:
+      LOG_INFO("Instruction access fault handler");
+      instruction_access_fault = true;
+      fault_handler_testutils_set_return_addr(&ret_addr);
+      break;
+    case kInstIllegalFault:
+      handler_instr_ill_fault();
+      break;
+    case kBkpt:
+      handler_bkpt();
+      break;
+    case kLoadAccFault:
+      handler_lsu_fault();
+      break;
+    case kStrAccFault:
+      handler_lsu_fault();
+      break;
+    case kECall:
+      handler_ecall();
+      break;
+    default:
+      while (1) {
```

Comment:
`while (true) { wait_for_interrupt(); }`

Infinite loops like this are UB.

Reply:
Done (the whole handler was just copy paste from `handlers.{h, c}`).